File: Reclaimer.Core/Geometry/Material.cs
using Reclaimer.Drawing;
using Reclaimer.Utilities;
using System.Diagnostics;
using System.Drawing;
using System.Numerics;

namespace Reclaimer.Geometry
{
    [DebuggerDisplay($"{{{nameof(Name)},nq}}")]
    public class Material
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public string AlphaMode { get; set; }
        public List<TextureMapping> TextureMappings { get; } = new();
        public List<MaterialTint> Tints { get; } = new();

        [Obsolete("legacy")]
        public int Flags { get; set; }
    }

    [DebuggerDisplay($"{{{nameof(GetDebuggerDisplay)}(),nq}}")]
    public class TextureMapping
    {
        public string Usage { get; set; }
        public ChannelMask BlendChannel { get; set; }
        public Texture Texture { get; set; }
        public ChannelMask ChannelMask { get; set; }
        public Vector2 Tiling { get; set; } = Vector2.One;

        private string GetDebuggerDisplay()
        {
            var result = $"[{Usage}] {{{Texture?.Name ?? "null"}}}";
            if (BlendChannel > 0)
                result = $"[{BlendChannel.ToString()[0]}] {result}";

            return result;
        }
    }

    [DebuggerDisplay($"{{{nameof(Name)},nq}}")]
    public class Texture
    {
        public int Id { get; set; }
        public float Gamma { get; set; } = 2.2f; //default to sRGB
        public IContentProvider<IBitmap> ContentProvider { get; set; }
        public string Name => ContentProvider?.Name;
        public DdsImage GetDds() => ContentProvider?.GetContent().ToDds(0);
    }

    public class MaterialTint
    {
        public string Usage { get; set; }
        public ChannelMask BlendChannel { get; set; }
        public Color Color { get; set; }
    }

    public static class AlphaMode
    {
        public const string Opaque = "opaque";
        public const string Clip = "clip";
        public const string Add = "additive";
        public const string Multiply = "multiply";
        public const string Blend = "blend";
        public const string PreMultiplied = "pre_multiplied";
    }

    public static class TextureUsage
    {
        public const string Other = "other";
        public const string BlendMap = "blend";
        public const string Diffuse = "diffuse";
        public const string DiffuseDetail = "diffuse_detail";
        public const string ColorChange = "color_change";
        public const string Normal = "bump";
        public const string NormalDetail = "bump_detail";
        public const string SelfIllumination = "self_illum";
        public const string Specular = "specular";
        public const string Transparency = "transparency";
    }

    public static class TintUsage
    {
        public const string Albedo = "albedo";
        public const string SelfIllumination = "self_illum";
        public const string Specular = "specular";
    }

    [Flags]
    public enum ChannelMask
    {
        Default = 0,

        Red = 1 << 0,
        Green = 1 << 1,
        Blue = 1 << 2,
        Alpha = 1 << 3,

        NonTransparent = Red | Green | Blue
    }
}

File: Reclaimer.Integration3d/reclaimer/src/Material.py
from dataclasses import dataclass
from typing import List
from enum import IntFlag

from .Types import *

__all__ = [
    'GAMMA_PRESET',
    'ALPHA_MODE',
    'TEXTURE_USAGE',
    'TINT_USAGE',
    'ChannelFlags',
    'Material',
    'TextureMapping',
    'TintColor',
    'Texture'
]


class GAMMA_PRESET:
    LINEAR = 1.0
    XRGB = 1.95
    SRGB = 2.2


class ALPHA_MODE:
    OPAQUE: str = 'opaque'
    CLIP: str = 'clip'
    ADD: str = 'additive'
    MULTIPLY: str = 'multiply'
    BLEND: str = 'blend'
    PRE_MULTIPLIED: str = 'pre_multiplied'


class TEXTURE_USAGE:
    BLEND: str = 'blend'
    DIFFUSE: str = 'diffuse'
    NORMAL: str = 'normal'
    HEIGHT: str = 'height'
    EMISSION: str = 'self_illum'
    SPECULAR: str = 'specular'
    TRANSPARENCY: str = 'transparency'
    COLOR_CHANGE: str = 'color_change'


class TINT_USAGE:
    ALBEDO = "albedo";
    EMISSION = "self_illum";
    SPECULAR = "specular";


class ChannelFlags(IntFlag):
    DEFAULT = 0

    RED = 1 << 0
    GREEN = 1 << 1
    BLUE = 1 << 2
    ALPHA = 1 << 3

    RGB = RED | GREEN | BLUE
    RGBA = RGB | ALPHA

    @staticmethod
    def get_default(usage: str) -> 'ChannelFlags':
        if usage == TEXTURE_USAGE.TRANSPARENCY:
            return ChannelFlags.ALPHA
        elif usage == TEXTURE_USAGE.BLEND:
            return ChannelFlags.RGBA
        else:
            return ChannelFlags.RGB


class Material(INamed, ICustomProperties):
    alpha_mode: str = None
    texture_mappings: List['TextureMapping']
    tints: List['TintColor']


@dataclass
class TextureMapping:
    texture_usage: str = None
    blend_channel: ChannelFlags = ChannelFlags.DEFAULT
    texture_index: int = -1
    channel_mask: ChannelFlags = ChannelFlags.DEFAULT
    tiling: Float2 = None


@dataclass
class TintColor:
    tint_usage: str = None
    blend_channel: ChannelFlags = ChannelFlags.DEFAULT
    tint_color: Color = None

class Texture(INamed, ICustomProperties):
    gamma: float = 2.2
    address: int = 0
    size: int = 0
